### PR TITLE
Merge 5.2.0minus 3

### PR DESCRIPTION
**Pull request by liam923**
This PR merges compiler changes to bring Merlin up to date with 5.2.0minus-3. This was a complicated merge for a number of reasons:
- It is the first merge since the compiler changed its directory layout
- There were a lot of conflicts due to the removal of jane syntax
- There was a compiler change that caused a dependency cycle in Merlin. More on that below.
- The addition of default modalities in signatures broke a lot of Merlin-specific code.

Regarding the dependency cycle: A compiler change made `utils/language_extension.ml(i)` depend on `parsing/location.ml(i)`. There are other files in `parsing/` that depend on files in `utils/`, so there is a mutual dependency between files in `parsing/` and `utils/`. While this compiles fine in the compiler, this dependency structure is problematic in Merlin because Merlin compiles each directory as an individual library. Rather than performing dune magic to resolve this, I opened [a PR on the compiler](https://github.com/ocaml-flambda/flambda-backend/pull/3269) that moves `utils/language_extension.ml(i)` to `parsing/language_extension.ml(i)`. The compiler version used for this merge is the tip of that PR rather than the tagged version.

Here's my review instructions:
- Review the first commit, which updates the import scripts to be able to handle the new directory structure. Actually, it's a bit worse than that. It needs to be able to handle two different directory structures, since for this merge, the old version uses the old structure and the new version uses the new structure.
- Gloss over the next three commits (3052ea2de25b8b37ebe75462fdc2511ec9e2ee52, bd98c3c9438e5155718d83e3bf86482ec95b1917, 2aa9ac6131abe4228eabf455687159ffbd1cf295), which import compiler changes and commit conflicts. I would usually say to skip them, except that this is the first run of the import script since the changes. Thus, it's worth doing some sort of spot-check.
- Review the remaining commits.

Here's some notes about some of the commits:
- 50305c20ce98149dcdffbb1157c55a856d156000 (Remove files that were deleted in flambda): The import script doesn't handle file removals or moves well (I've created tickets internally to fix this). This manually deals with some such files.
- 49edf67feb67dce014ebe6ee0b61624d45dc9000 (Apply old diff to language_extension): The `Language_extension` module moved locations as a result of this merge (see above note about dependency cycle). This manually ports over some Merlin-specific changes to that file.
- e28fc5223b98e4a75590b475fb85e3b77835fe09 (Update magic number script): The compiler changed how it handles magic numbers since the last merge. This commit updates the script that bumps magic numbers to reflect this.

### CODE REVIEW

**Comment by goldfirere** [comment]
I’m very much on the back foot right now. I nominate @dkalinichenko-js to review. I know she does not have (much?) experience in merlin, but I think the nature of the changes here does not require that much experience. Diana, if there are parts that want further review, please highlight and I can take a look. Thanks!

**Comment by dkalinichenko-js** [comment]
The `subdirectory` bits in import scripts will be unnecessary after this is merged, right? Should we leave a CR to clean them up?

**Comment by liam923** [comment]
@dkalinichenko-js yes, but imo it's best to keep it around incase we ever do a change in the future. I'll leave a comment saying something to this effect.

**Comment by dkalinichenko-js** [comment]
@goldfirere could you look at https://github.com/janestreet/merlin-jst/pull/119/commits/2d99ac1ace70fcef55e32dcc731f838248a534ad, especially the `combine_sigs` function? Everything else look good to me.

File: src/analysis/ptyp_of_type.ml
Comment (goldfirere) [inline src/analysis/ptyp_of_type.ml:38]: Seems easy enough to do better -- but maybe the point is that we're not sure what we want to display to users? (This doesn't need to hold up the merge.)

File: src/analysis/ptyp_of_type.ml
Comment (goldfirere) [inline src/analysis/ptyp_of_type.ml:280]: Use `Ast_helper.Sg.mk` (here and elsewhere)

**Review by goldfirere — APPROVED** [review]
Quick review, focusing on `combine_sigs`. Some style suggestions, but all looks correct.

File: src/analysis/ptyp_of_type.ml
Comment (liam923) [reply src/analysis/ptyp_of_type.ml:38]: I'm not sure what we want to display to the user, and I'd rather print the jkind to little than too often (this is only used for construct), so I think it's best to just not include the kind for now. I also suspect this might be easier to do once https://github.com/ocaml-flambda/flambda-backend/pull/3224 is merged.

**Pull request merged**
